189 Player Selection V1

Straczi opened the pull request:
[#189](https://github.com/corewar-teamprojekt/corewar/issues/189)

0815Sailsman commented on frontend/app/src/domain/LobbyStatus.ts:
```diff
@@ -0,0 +1,10 @@
+	playerBSubmitted: boolean;
+	gameState: GameState;
+	result: {
+		winner: "A" | "b" | "DRAW";
```
'b' should be lowercase

0815Sailsman commented on frontend/app/src/services/rest/LobbyRest.ts:
id like a v1 in the name of all functions accessing v1 endpoints for consistency

0815Sailsman commented on frontend/app/src/services/rest/LobbyRest.ts:
```diff
@@ -0,0 +1,50 @@
+	}
+}
+
+export async function createLobby(): Promise<number> {
+	return await fetch(`${import.meta.env.VITE_REACT_APP_BACKEND_URL}/v1/lobby`, {
+		method: "POST",
+	}).then((response) => response.json().then((data) => data.lobbyId as number));
+}
```
this needs a body containing a playername

0815Sailsman replied:
https://github.com/corewar-teamprojekt/corewar/blob/main/api-v1.md#post-lobby

0815Sailsman commented on frontend/app/src/pages/playerSelection/PlayerSelection.tsx:
```diff
@@ -1,49 +1,139 @@
+				.then(() => {
+					if (!lobby || lobby.id !== lobbyId) {
+						const newLobby = new Lobby(
+							lobbyId,
+							[userName],
+							GameState.NOT_STARTED,
+						);
+						lobbyDispatch({ type: "join", lobby: newLobby });
+					}
+					setShouldRedirect(true);
+				});
+		}
 
-	function setPlayerAndRedirectToCoding(
-		actionType: "setPlayerA" | "setPlayerB",
-	) {
-		if (dispatcher) {
+		function handleLobbyError() {
+			toast({
+				title: "Oopsie👉👈",
+				description: "Something went wrong while joining the lobby 😢",
+				variant: "destructive",
+			});
+			if (userDispatcher) userDispatcher({ type: "logout", user: null });
+			navigate("/lobby-selection");
+			setBlockLogout(false);
+		}
+
+		if (!user || !lobbyDispatch) {
+			return;
+		} else if (lobby) {
+			joinLobbyWithPlayer(user.name, lobby.id);
+		} else {
+			createLobby()
+				.catch(handleLobbyError)
+				.then((lobbyID) => {
+					if (lobbyID) joinLobbyWithPlayer(user.name, lobbyID);
+				});
+		}
+	}, [user, lobby, lobbyDispatch, toast, userDispatcher, navigate]);
```
does this have to be a single useEffectHook? Seems very long :/

Straczi replied:
the logic of this one is actually not that long (the if-else at the end) and if i would put the two functions outside of the useEffect i would get ESLINT warnings saying to update the dependency array. (the official fix for this is to declare the function inside the useEffect hooks : [see here](https://github.com/facebook/react/issues/14920)

Straczi replied:
but i tried to reduce the visual complexity a bit by reducing the use of .then(....) and using await statements instead

0815Sailsman commented on frontend/app/src/pages/playerSelection/PlayerSelection.tsx:
```diff
@@ -1,49 +1,139 @@
+		} else {
+			createLobby()
+				.catch(handleLobbyError)
+				.then((lobbyID) => {
+					if (lobbyID) joinLobbyWithPlayer(user.name, lobbyID);
+				});
```
the player who created the lobby is already joined. no need to join again. this will even fail, because that player is already in the lobby (needs changes in create lobby call, see spec)

Straczi replied:
aight :+1:

0815Sailsman commented on frontend/app/src/pages/playerSelection/PlayerSelection.test.tsx:
i'll read this entire thing, once create lobby has been adjusted, but seems good on first glance

0815Sailsman reviewed (changes requested):
it feels a little weird, that the scale happens so early, because the hitboxes of the icons are a lot bigger than the hexagons themselves. but I think its good as is and not a prio atm, possibly a story for later.

also, you already added parts of the routing logic of the final lobby system. in this state, v0 is broken. we should either re-enable v0 functionality and disable v1 logic for now, or switch entirely. (you wouldn't need to delete disabled stuff, but rather hide it in a comment or unused function for now if we decide to go down that route)

Straczi commented:
> it feels a little weird, that the scale happens so early, because the hitboxes of the icons are a lot bigger than the hexagons themselves. but I think its good as is and not a prio atm, possibly a story for later.
> 
> also, you already added parts of the routing logic of the final lobby system. in this state, v0 is broken. we should either re-enable v0 functionality and disable v1 logic for now, or switch entirely. (you wouldn't need to delete disabled stuff, but rather hide it in a comment or unused function for now if we decide to go down that route)

i don’t know if i am a fan  of that, since this way v1 becomes a bit shitty, since you never know which player already is in the lobby and which one is free to pick? (And the ticket is litteraly named player-selection-v1, i did not know, we should also support v0, that would be nice to be mentioned in the issue)

Straczi commented:
> also, you already added parts of the routing logic of the final lobby system. in this state, v0 is broken. we should either re-enable v0 functionality and disable v1 logic for now, or switch entirely. (you wouldn't need to delete disabled stuff, but rather hide it in a comment or unused function for now if we decide to go down that route)

since the ticket is named player selection v1 i implemented it with v1 functionality.  If i implemented it differently it would not be v1 anymore but v0 and would not fit the issue right? Also what parts do you mean exactly? Would like to talk about this (maybe in daily :) ), that would be easier than communicating via PR comments i think :sweat_smile:

0815Sailsman commented:
As discussed in the daily, this will be merged together with #270 and is pretty much done

0815Sailsman reviewed (approved):
As discussed in the daily, this will be merged together with #270 and is pretty much done

Straczi commented:
merged via #287
